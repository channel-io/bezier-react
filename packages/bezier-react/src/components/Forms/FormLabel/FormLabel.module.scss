.LabelText {
  display: block;
  word-break: break-word;
}
