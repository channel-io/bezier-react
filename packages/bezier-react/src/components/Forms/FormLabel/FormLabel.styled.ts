/* Internal dependencies */
import { styled } from 'Foundation'
import { Tooltip as BaseTooltip } from 'Components/Tooltip'
import { Text } from 'Components/Text'
import { Icon } from 'Components/Icon'

export const Label = styled(Text)`
  display: block;
  text-align: left;
`

export const Box = styled.div`
  display: flex;
  align-items: center;
`

export const HelpIcon = styled(Icon)``

export const Tooltip = styled(BaseTooltip)`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: 6px;

  &:hover {
    > ${HelpIcon} {
      color: ${({ foundation }) => foundation?.theme?.['txt-black-darkest']};
    }
  }
`
