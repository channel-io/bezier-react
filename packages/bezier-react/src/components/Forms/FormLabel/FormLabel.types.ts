/* Internal dependencies */
import { BezierComponentProps, ChildrenProps, IdentifierProps } from 'Types/ComponentProps'
import { TextProps } from 'Components/Text'

interface FormLabelOptions {
  htmlFor?: string
  help?: React.ReactNode
}

export default interface FormLabelProps extends
  BezierComponentProps,
  ChildrenProps,
  TextProps,
  Partial<IdentifierProps>,
  FormLabelOptions {}
