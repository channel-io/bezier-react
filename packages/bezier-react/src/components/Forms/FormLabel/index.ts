export { default as FormLabel } from './FormLabel'
export type { default as FormLabelProps } from './FormLabel.types'
