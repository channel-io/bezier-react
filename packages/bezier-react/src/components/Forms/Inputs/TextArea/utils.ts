import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
