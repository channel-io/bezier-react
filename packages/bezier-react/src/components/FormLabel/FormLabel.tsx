import React, { forwardRef } from 'react'

import classNames from 'classnames'

import { isEmpty } from '~/src/utils/type'

import { useFormControlContext } from '~/src/components/FormControl'
import { Help } from '~/src/components/Help'
import { HELP_DISPLAY_NAME } from '~/src/components/Help/Help'
import {
  LegacyHStack,
  LegacyStackItem,
} from '~/src/components/LegacyStack'
import { Text } from '~/src/components/Text'

import { type FormLabelProps } from './FormLabel.types'

import styles from './FormLabel.module.scss'

export const FORM_LABEL_TEST_ID = 'bezier-form-label'

/**
 * `FormLabel` is a component to show label.
 * `FormControl` component can handle its layout by `position` props.
 * @example
 * ```tsx
 * <FormControl position="top">
 *   <FormLabel>
 *     Name
 *   </FormLabel>
 *   <TextField />
 * </FormControl>
 * ```
 */
export const FormLabel = forwardRef<HTMLLabelElement, FormLabelProps>(function FormLabel(props, forwardedRef) {
  const {
    help,
    bold = true,
    color = 'txt-black-darkest',
    children,
    ...rest
  } = props

  const contextValue = useFormControlContext()
  const {
    className: formControlClassName,
    ...ownProps
  } = contextValue?.getLabelProps(rest) ?? {
    typo: '13',
    className: undefined,
    ...rest,
  }

  const HelpComponent = (() => {
    if (isEmpty(help)) { return null }

    if (React.isValidElement(help)) {
      // FIXME(@bora): string, JSXElementCostructor 타입에 displayName 속성이 존재하지 않음, 임시로 any 타입으로 설정
      const { displayName } = help?.type as any
      if (displayName === HELP_DISPLAY_NAME) {
        return help
      }
    }

    return (
      <Help>
        { help }
      </Help>
    )
  })()

  const LabelComponent = (
    <Text
      className={classNames(
        styles.LabelText,
        !HelpComponent && formControlClassName,
      )}
      ref={forwardedRef}
      as="label"
      align="left"
      bold={bold}
      color={color}
      data-testid={FORM_LABEL_TEST_ID}
      {...ownProps}
    >
      { children }
    </Text>
  )

  if (isEmpty(children)) { return null }

  return (
    !HelpComponent
      ? LabelComponent
      : (
        <LegacyHStack align="center" spacing={6} className={HelpComponent && formControlClassName}>
          <LegacyStackItem shrink weight={1}>
            { LabelComponent }
          </LegacyStackItem>
          <LegacyStackItem>
            { HelpComponent }
          </LegacyStackItem>
        </LegacyHStack>
      ))
})
