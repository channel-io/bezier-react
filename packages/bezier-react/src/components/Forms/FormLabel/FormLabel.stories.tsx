/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import type { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from '~/src/utils/storyUtils'
import FormLabel from './FormLabel'
import FormLabelProps from './FormLabel.types'

export default {
  title: getTitle(base),
  component: FormLabel,
  argTypes: {
    children: {
      control: {
        type: 'text',
      },
    },
    help: {
      control: {
        type: 'text',
      },
    },
  },
} as Meta

const Template: Story<FormLabelProps> = props => <FormLabel {...props} />

export const Primary: Story<FormLabelProps> = Template.bind({})
Primary.args = {
  htmlFor: 'test',
  help: 'Lorem ipsum',
  children: 'Label',
}
