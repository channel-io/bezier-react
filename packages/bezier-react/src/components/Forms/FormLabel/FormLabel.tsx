/* External dependencies */
import React, { forwardRef, useMemo } from 'react'

/* Internal dependencies */
import { Typography } from 'Foundation'
import { isEmpty } from 'Utils/typeUtils'
import useFormControlContext from 'Components/Forms/useFormControlContext'
import { Help } from 'Components/Help'
import { HELP_DISPLAY_NAME } from 'Components/Help/Help'
import { HStack, StackItem } from 'Components/Stack'
import type FormLabelProps from './FormLabel.types'
import * as Styled from './FormLabel.styled'

export const FORM_LABEL_TEST_ID = 'bezier-react-form-label'

function FormLabel({
  testId = FORM_LABEL_TEST_ID,
  help,
  as = 'label',
  bold = true,
  color = 'txt-black-darkest',
  children,
  ...rest
}: FormLabelProps,
forwardedRef: React.Ref<HTMLLabelElement>,
) {
  const contextValue = useFormControlContext()

  const { Wrapper, typo, ...ownProps } = contextValue?.getLabelProps(rest) ?? {
    Wrapper: React.Fragment,
    typo: Typography.Size13,
    ...rest,
  }

  const LabelComponent = useMemo(() => (
    <Styled.Label
      {...ownProps}
      ref={forwardedRef}
      testId={testId}
      forwardedAs={as}
      bold={bold}
      typo={typo}
      color={color}
    >
      { children }
    </Styled.Label>
  ), [
    as,
    typo,
    bold,
    color,
    testId,
    children,
    forwardedRef,
    ownProps,
  ])

  const HelpComponent = useMemo(() => {
    if (isEmpty(help)) { return null }

    if (React.isValidElement(help)) {
      // FIXME(@bora): string, JSXElementCostructor 타입에 displayName 속성이 존재하지 않음, 임시로 any 타입으로 설정
      const { displayName } = help?.type as any
      if (displayName === HELP_DISPLAY_NAME) {
        return help
      }
    }

    return (
      <Help>
        { help }
      </Help>
    )
  }, [help])

  if (isEmpty(children)) { return null }

  return (
    <Wrapper>
      { !HelpComponent
        ? LabelComponent
        : (
          <HStack align="center" spacing={6}>
            <StackItem shrink weight={1}>
              { LabelComponent }
            </StackItem>
            <StackItem>
              { HelpComponent }
            </StackItem>
          </HStack>
        ) }
    </Wrapper>
  )
}

export default forwardRef(FormLabel)
