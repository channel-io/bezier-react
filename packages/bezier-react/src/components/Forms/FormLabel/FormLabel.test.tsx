import React from 'react'

import { render } from '~/src/utils/test'

import Help, { HELP_TEST_ID } from '~/src/components/Help/Help'

import FormLabel, { FORM_LABEL_TEST_ID } from './FormLabel'
import type FormLabelProps from './FormLabel.types'

describe('FormLabel >', () => {
  let props: FormLabelProps

  beforeEach(() => {
    props = {
      help: '',
      children: 'label',
    }
  })

  const renderComponent = (otherProps?: Partial<FormLabelProps>) =>
    render(<FormLabel {...props} {...otherProps} />)

  it('Snapshot >', () => {
    const { getByTestId } = renderComponent()
    const rendered = getByTestId(FORM_LABEL_TEST_ID)

    expect(rendered).toMatchSnapshot()
  })

  it('renders text when children prop is not empty', () => {
    const { getByTestId } = renderComponent()
    const rendered = getByTestId(FORM_LABEL_TEST_ID)

    expect(rendered).toBeInTheDocument()
  })

  it('renders nothing when children prop is empty', () => {
    const { queryByTestId } = renderComponent({ children: '' })
    const rendered = queryByTestId(FORM_LABEL_TEST_ID)

    expect(rendered).toBeNull()
  })

  it('renders help icon when help prop is not empty', () => {
    const { getByTestId } = renderComponent({ help: 'test' })
    const rendered = getByTestId(HELP_TEST_ID)

    expect(rendered).toBeInTheDocument()
  })

  it('renders help component when HelpTooltip is not empty', () => {
    const { getByTestId } = renderComponent({
      help: <Help> test </Help>,
    })
    const rendered = getByTestId(HELP_TEST_ID)

    expect(rendered).toBeInTheDocument()
  })
})
