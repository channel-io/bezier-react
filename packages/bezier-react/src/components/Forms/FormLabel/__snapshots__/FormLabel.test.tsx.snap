// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLabel > Snapshot > 1`] = `
.c0 {
  display: block;
  text-align: left;
  word-break: break-word;
}

<label
  class="margin Text typo-13 bold c0"
  data-testid="bezier-react-form-label"
  foundation="[object Object]"
  style="--b-text-color: var(--txt-black-darkest);"
>
  label
</label>
`;
